source/material_model/melt_global.cc

```diff
@@ -157,16 +157,19 @@ namespace aspect
                                    0.0;
           out.densities[i] = (reference_rho_s + delta_rho) * temperature_dependence
                              * std::exp(compressibility * (in.pressure[i] - this->get_surface_pressure()));
+           
+	  AssertThrow(this->introspection().compositional_name_exists("porosity"),
```

Comment:
This may break some things, because the Assert is now executed even if `this->include_melt_transport() == false` and in this case the model may not have a porosity field. 

So what I would suggest is to split this up and put the logic back into the loop, but change the condition to 
`if (this->include_melt_transport())`, and then have a separate if-statement before line 175 (`// Calculate the melting rate...`) saying `if (include_melting_and_freezing && in.strain_rate.size())`. 
Then you can compute the viscosity as `out.viscosities[i] = eta_0;` outside of the loop, `out.viscosities[i] *= exp(- alpha_phi * porosity);` for models with melt transport, and additionally multiply with `out.viscosities[i] *= depletion_strengthening;` for models that also have melting and freezing. 

Sorry for not noticing this earlier.

Reply:
Since it's a "global melt", it would makes sense to require a porosity field to exist anyway, no? But maybe then it would needs to go somewhere else?

Reply:
OK, we actually need it for one cookbook! So let's keep it. I'm trying to modify to put the different if statements.